rfcs/0032-automated-validation.md

```diff
@@ -0,0 +1,229 @@
+
+The following diagram shows the runtime architecture of the processes in the system once initialization has been completed. Note how this diagram shows that statistics may consume multiple providers, and validations may consume multiple statistics.
+
+![](./res/automated_validation_runtime_architecture.conv.tex.png)
+
+#### Resources
+[detailed-design-architecture-resources]: #detailed-design-architecture-resources
+
+A deployment consists of a series of pods and containers. We define a universal abstraction for reasoning about these artifacts of a deployment, which generically call "resources". Resources are classified into different categories in an subtyping model. For example, the main resource we are interested in is a `CodaNode`. There are multiple subclasses of `CodaNode`, such as `BlockProducer`, `SnarkCoordinator`, or `Seed`. Each of these classes may have different metadata associate with instances of them. For instance, all `CodaNode` instances have a `pod_name`, and `BlockProducer` instances have additional metadata for what block producer role they are and their integer id (eg in "fish-block-producer-1", role="fish" and id=1).
+
+Resources can be collected into queriable datasets which allow selection of resources by metadata filters or classification.
+
+#### Providers
+[detailed-design-architecture-providers]: #detailed-design-architecture-providers
+
+Providers provide network state to the rest of the system. There are multiple kinds of providers, such as `GraphQLProviders`, which provide data from GraphQL queries on a periodic basis, and `EventProviders`/`LogProviders`, which extract events from logs. Logs can be streamed back to the automated validation service fairly easily from StackDriver by creating a log sink, pub sub topic, and one or more filter subscriptions in Google Cloud (I have tested this functionality for pull-based subscriptions, and push-based subscriptions are an easy extension of this to setup). Providers are dispatched for specific resources, and provide an internal subscription to state updates for those resources so that statistics may receive updates.
+
+#### Statistics
+[detailed-design-architecture-statistics]: #detailed-design-architecture-statistics
+
+Statistics consume network state information from one or more providers and incrementally compute a rolling state (representing the computed statistic). Statistics are (at least in the initial design) individually scoped to resources (meaning that each statistic has a singular instance computed for each resource that is being monitored by the system). Statistics broadcast their state updates so that vlaidations can be performed on one or more statistic as values change. Statistics can also be forwarded to as prometheus metrics. Doing this would allow us to compute more meaningful network wide statistics that we cannot easily compute today without writing a program to scrape the logs after the network runs.
```

Comment:
https://joshday.github.io/OnlineStats.jl/v0.10/pages/types.html we can algorithms from this as we need them I guess. also, typo, vlaidations